xline/src/server/maintenance.rs

```diff
@@ -0,0 +1,225 @@
+        DowngradeResponse, HashKvRequest, HashKvResponse, HashRequest, HashResponse, Maintenance,
+        MoveLeaderRequest, MoveLeaderResponse, SnapshotRequest, SnapshotResponse, StatusRequest,
+        StatusResponse,
+    },
+    storage::storage_api::StorageApi,
+};
+
+/// Snapshot chunk size
+const SNAPSHOT_CHUNK_SIZE: u64 = 64 * 1024;
+
+/// Maintenance Server
+#[derive(Debug)]
+pub(crate) struct MaintenanceServer<S>
+where
+    S: StorageApi,
+{
+    /// persistent storage
+    persistent: Arc<S>,
+    /// Header generator
+    header_gen: Arc<HeaderGenerator>,
+}
+
+impl<S> MaintenanceServer<S>
+where
+    S: StorageApi,
+{
+    /// New `LeaseServer`
+    pub(crate) fn new(persistent: Arc<S>, header_gen: Arc<HeaderGenerator>) -> Self {
+        Self {
+            persistent,
+            header_gen,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<S> Maintenance for MaintenanceServer<S>
+where
+    S: StorageApi,
+{
+    async fn alarm(
+        &self,
+        _request: tonic::Request<AlarmRequest>,
+    ) -> Result<tonic::Response<AlarmResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "alarm is unimplemented".to_owned(),
+        ))
+    }
+
+    async fn status(
+        &self,
+        _request: tonic::Request<StatusRequest>,
+    ) -> Result<tonic::Response<StatusResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "status is unimplemented".to_owned(),
+        ))
+    }
+
+    async fn defragment(
+        &self,
+        _request: tonic::Request<DefragmentRequest>,
+    ) -> Result<tonic::Response<DefragmentResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "defragment is unimplemented".to_owned(),
+        ))
+    }
+
+    async fn hash(
+        &self,
+        _request: tonic::Request<HashRequest>,
+    ) -> Result<tonic::Response<HashResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "hash is unimplemented".to_owned(),
+        ))
+    }
+
+    async fn hash_kv(
+        &self,
+        _request: tonic::Request<HashKvRequest>,
+    ) -> Result<tonic::Response<HashKvResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "hash_kv is unimplemented".to_owned(),
+        ))
+    }
+
+    type SnapshotStream = ReceiverStream<Result<SnapshotResponse, tonic::Status>>;
+
+    async fn snapshot(
+        &self,
+        _request: tonic::Request<SnapshotRequest>,
+    ) -> Result<tonic::Response<Self::SnapshotStream>, tonic::Status> {
+        let mut snapshot = self.persistent.get_snapshot().map_err(|e| {
+            error!("get snapshot failed, {e}");
+            tonic::Status::internal("get snapshot failed")
+        })?;
+
+        let header = self.header_gen.gen_header();
+        let (tx, rx) = mpsc::channel(1);
+        let _ig = tokio::spawn(async move {
+            if let Err(e) = snapshot.rewind() {
+                error!("snapshot seek failed, {e}");
```

Comment:
I think "snapshot rewind failed" is more straightforward.

Reply:
Fixed